pkg/assembler/grammar.go

```diff
@@ -0,0 +1,199 @@
+package assembler
+
+import (
+	"fmt"
+	"math"
+)
+
+// Grammar and AST
+
+type CasmProgram struct {
+	Instructions []Instruction `@@`
+}
+
+type Instruction struct {
+	Core      *CoreInstruction `@@`
+	ApPlusOne bool             `( "," @"ap" "+" "+" )? ";" |`
+	ApPlus    *ApPlus          `@@ ";"`
+}
+
+type CoreInstruction struct {
+	AssertEq *AssertEq `@@ |`
+	Jnz      *Jnz      `@@ |`
+	Jump     *Jump     `@@ |`
+	Call     *Call     `@@ |`
+	Ret      *Ret      `@@ `
+}
+
+type AssertEq struct {
+	Dst   *Deref      `@@`
+	Value *Expression `"=" @@`
+}
+
+type Jump struct {
+	JumpType string      `"jmp" @("rel" | "abs")`
+	Value    *Expression `@@`
+}
+
+type Jnz struct {
+	Value     *DerefOrImm `"jmp" "rel" @@`
+	Condition *Deref      `"if" @@ "!" "=" "0"`
+}
+
+type Call struct {
+	CallType string      `"call" @("rel" | "abs")`
+	Value    *DerefOrImm `@@`
+}
+
+type Ret struct {
+	Ret string `"ret"`
+}
+
+type ApPlus struct {
+	Value *Expression `"ap" "+" "=" @@`
+}
+
+type Expression struct {
+	DoubleDeref   *DoubleDeref   `@@ |`
+	MathOperation *MathOperation `@@ |`
+	Deref         *Deref         `@@ |`
+	Immediate     *string        `@Int`
+}
+
+type Deref struct {
+	Name   string  `"[" @("ap" | "fp")`
+	Offset *Offset `@@? "]"`
+}
+
+type DoubleDeref struct {
+	Deref  *Deref  `"[" @@`
+	Offset *Offset `@@? "]"`
+}
+
+type Offset struct {
+	Sign  string `@("+" | "-")`
+	Value *int   `@Int`
+}
+
+type MathOperation struct {
+	Lhs      *Deref      `@@`
+	Operator string      `@("+" | "*")`
+	Rhs      *DerefOrImm `@@`
```

Comment:
The whitepaper gives the following cases for `Rhs`:

* Immediate
* `Deref`
* `MathOperation` (with `Deref`s for lhs and rhs)
* `MathOperation` (with `Deref` for lhs and Immediate for rhs)
* `DoubleDeref`

Reply:
Yep, the Rhs for `AssertEq`. For `MathOperation`, the right hand side is only Deref and Immediate.